crates/trigger-http/src/server.rs

```diff
@@ -138,12 +148,46 @@ impl<F: RuntimeFactors> HttpServer<F> {
 
     /// Serve incoming requests over the provided [`TcpListener`].
     pub async fn serve(self: Arc<Self>) -> anyhow::Result<()> {
-        let listener = TcpListener::bind(self.listen_addr).await.with_context(|| {
-            format!(
-                "Unable to listen on {listen_addr}",
-                listen_addr = self.listen_addr
-            )
-        })?;
+        let listener: TcpListener = match TcpListener::bind(self.listen_addr).await {
+            Ok(listener) => listener,
+            Err(err) => {
+                if self.find_free_port && err.kind() == ErrorKind::AddrInUse {
+                    let mut found_listener = None;
+                    let mut addr = self.listen_addr;
+                    for _ in 1..=9 {
+                        if addr.port() == u16::MAX {
+                            anyhow::bail!(
+                                "Couldn't find a free port as we've reached the maximum port number. Consider retrying with a lower base port."
+                            );
+                        }
+
+                        addr.set_port(addr.port() + 1);
+
+                        match TcpListener::bind(addr).await {
+                            Ok(listener) => {
+                                found_listener = Some(listener);
+                                break;
+                            }
+                            Err(err) => {
+                                if err.kind() == ErrorKind::AddrInUse {
+                                    continue;
+                                }
+                                anyhow::bail!("Unable to listen on {}", self.listen_addr);
+                            }
```

Comment:
We probably shouldn't just drop the original error in here. Separately, you can move the `if/continue` out to a match guard clause:

```suggestion
                            Err(err) if err.kind() != ErrorKind::AddrInUse => {
                            	// Try the next port
                            }
                            Err(err) => {
                                anyhow::bail!("Unable to listen on {addr}: {err:?}");
                            }
                            
```